### ISSUE TITLE
[python-package] LightGBM outputs validation error even when validation target is entirely NaN

### ISSUE THREAD

**Issue by bceran** [microsoft/LightGBM#7041]
When using LightGBM with a validation set where all target labels are NaN, the training process still calculates the validation loss (e.g., L2). 

Here is a code sample:

```python
import numpy as np
import pandas as pd
import lightgbm as lgb

feature_dict = {}

for i in range(15):
	feature_dict[f"feature_{i}"] = np.random.rand(10000)
	
df_train = pd.DataFrame(feature_dict)
df_val = pd.DataFrame(feature_dict)

df_train["target"] = np.random.rand(10000)
df_val["target"] = [np.nan for _ in range(10000)]  # all validation targets -> NaN

params = {'boosting_type': 'gbdt'}

train_data = lgb.Dataset(df_train.drop(['target'], axis=1),
						 label=df_train['target'])
						 
valid_data = lgb.Dataset(df_val.drop(['target'], axis=1),
					   label=df_val['target'])

callbacks = [lgb.log_evaluation(period=20)]

model = lgb.train(params, train_data, valid_sets=[train_data, valid_data], callbacks=callbacks)
```

my output:

```text
[20]	training's l2: 0.0757915	valid_1's l2: 0.252605
[40]	training's l2: 0.0702904	valid_1's l2: 0.253532
[60]	training's l2: 0.0653011	valid_1's l2: 0.254637
[80]	training's l2: 0.060823	valid_1's l2: 0.255788
[100] training's l2: 0.0565347	valid_1's l2: 0.257031
```

lightgbm version : 4.6.0

**Comment by jameslamb** [comment]
Thanks for using LightGBM and taking the time to write up this repo.

I've reformatted your post a bit so that the different between code and your other text is clearer. If you're not sure how to do that, please review https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax

> _outputs validation error_
> ...
> _When using LightGBM with a validation set where all target labels are NaN, the training process still calculates the validation loss_

Ah ok, so by "validation error" you do not mean a Python exception, but you're just saying you think it is wrong that metrics could be calculated when the target is all NaN values?

That is intentional behavior... NaN values in labels are replaced by 0 in LightGBM. See that here:

https://github.com/microsoft/LightGBM/blob/f1d3181ced9fd01f4b2899054abd99be6773e939/src/io/metadata.cpp#L388

https://github.com/microsoft/LightGBM/blob/f1d3181ced9fd01f4b2899054abd99be6773e939/include/LightGBM/utils/common.h#L653-L655

Here's a minimal, reproducible example you can use to examine that:

```python
import numpy as np
import lightgbm as lgb
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split

X, y = make_regression(n_samples=1_000, n_features=4, n_informative=4, random_state=708)
X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.2, random_state=708)

params = {
    "objective": "regression",
    "num_leaves": 7,
    "deterministic": True,
    "force_row_wise": True,
    "seed": 773,
}

# case 1: target in validation set is all NaN values
y_valid_all_nan = np.full_like(y_valid, fill_value=np.nan)
model = lgb.train(
	params=params,
	train_set=lgb.Dataset(X_train, label=y_train),
	num_boost_round=10,
	valid_sets=[
		lgb.Dataset(X_valid, label=y_valid_all_nan)
	],
	callbacks=[lgb.log_evaluation(period=1)]
)

# case 2: target in validation set is all 0.0
y_valid_all_0 = np.full_like(y_valid, fill_value=0.0)
model = lgb.train(
	params=params,
	train_set=lgb.Dataset(X_train, label=y_train),
	num_boost_round=10,
	valid_sets=[
		lgb.Dataset(X_valid, label=y_valid_all_0)
	],
	callbacks=[lgb.log_evaluation(period=1)]
)
```

In both cases, the evaluation metrics are identical:

```text
[LightGBM] [Info] Total Bins 1020
[LightGBM] [Info] Number of data points in the train set: 800, number of used features: 4
[LightGBM] [Info] Start training from score 3.629214
[1]     valid_0's l2: 99.3556
[2]     valid_0's l2: 272.379
[3]     valid_0's l2: 546.412
[4]     valid_0's l2: 872.483
[5]     valid_0's l2: 1171.8
[6]     valid_0's l2: 1495.96
[7]     valid_0's l2: 1798.74
[8]     valid_0's l2: 2082.72
[9]     valid_0's l2: 2426.72
[10]    valid_0's l2: 2815.32
```

If you'd like to propose changing that behavior in LightGBM, please post about that here: #5592
